File: lessons/04_03/Golang/2_abstraction.go
package main

import "fmt"

type Animal struct {
	health     int
	speed      float64
	environment string
}

func (a *Animal) GetHealth() int {
	return a.health
}

func (a *Animal) SetHealth(health int) {
	a.health = health
}

func (a *Animal) GetSpeed() float64 {
	return a.speed
}

func (a *Animal) SetSpeed(speed float64) {
	a.speed = speed
}

func (a *Animal) AnimalInfo() {
	fmt.Println("Animal lives in", a.environment)
}

func main() {
	animals := map[string]Animal{
		"Pigeon":  {100, 40, "Air"},
		"Dog":     {90, 25, "Earth"},
		"Dolphin": {65, 70, "Water"},
	}

	for name, animal := range animals {
		fmt.Printf("\t%s INFO:\n", name)
		fmt.Printf("%s health: %d\n", name, animal.GetHealth())
		fmt.Printf("%s speed: %.2f\n", name, animal.GetSpeed())
		animal.AnimalInfo()
		fmt.Println()
	}
}

File: lessons/04_03/Java/2_abstraction.java
class Animal {
  private int health;
  private float speed;
  private String environment;

  public Animal(int health, float speed, String environment) {
    this.health = health;
    this.speed = speed;
    this.environment = environment;
  }

  public int getHealth() {
    return health;
  }

  public void setHealth(int health) {
    this.health = health;
  }

  public float getSpeed() {
    return speed;
  }

  public void setSpeed(float speed) {
    this.speed = speed;
  }

  public void animalInfo() {
    System.out.println("Animal lives in " + environment);
  }
}

public class Main {
    public static void main(String[] args) {
      Animal[] animals = {
        new Animal(100, 40, "Air"),
        new Animal(90, 25, "Earth"),
        new Animal(65, 70, "Water")
      };

      String[] names = {"Pigeon", "Dog", "Dolphin"};

      for (int i = 0; i < animals.length; i++) {
        System.out.println("\t" + names[i].toUpperCase() + " INFO:");
        System.out.println(names[i] + " health: " + animals[i].getHealth());
        System.out.println(names[i] + " speed: " + animals[i].getSpeed());
        animals[i].animalInfo();
        System.out.println();
      }
  }
}
